Add pwsh support

ACSimon33 opened the pull request:
Hi,
since there seemed to be no progress on the pwsh support ticket, I thought I'd just try it myself.

I've took the hints from the issue and addressed everything which is applicable to pwsh:

> - [x]  [renderFalse](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L370) procedure to output a false shell statement
> - [x]  [renderTrue](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L418) procedure to output a true shell statement
> - [x]  [renderText](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L458) procedure to echo text

> [renderSettings](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L5350) procedure to:
> - [x]  define environment variable
> - [x]  unset environment variable
> - [x]  define shell alias (if this concept applies to PowerShell)
> - [x]  unset shell alias
> - [x]  define shell function (if this concept applies to PowerShell)
> - [x]  unset shell function

Aliases are defined as functions as well because pwsh doesn't support aliases that contain whitespaces or consist of multiple commands that take arguments.

> - [ ]  execute xrdb command to define XResources
> - [ ]  execute xrdb command to unset XResources

The two above are not applicable to pwsh as far as I can tell (at least on Windows)

> - [x]  execute PowerShell change directory command
> - [x]  [renderAutoinit](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L5807) to output pwsh code to define the module function for the shell,

The autoinit works only on Unix platforms. On Windows, I defined the the environment and the  `envmodule` command directly in the init script.

> - [x]  pwsh should also be [added to the list of supported shells](https://github.com/cea-hpc/modules/blob/15158c3323edb3ab65a721ae8842346a3615c52b/modulecmd.tcl.in#L11018).


I've tested everything manually on Windows and Linux but I still need to get the testsuite running for pwsh. Also, I probably need to add something to the documentation. In the meantime, any feedback is appreciated.

Best,
Simon


Closes #326

xdelaruelle commented on tcl/envmngt.tcl.in:
```diff
@@ -1236,6 +1278,25 @@ ml <- function(...){
+         }
+
+         set fdef "function global:envmodule {"
+         append fdef {
```
Support for `MODULES_REDIRECT_OUTPUT` should be added (output sent to stderr or stdout)

xdelaruelle commented on tcl/envmngt.tcl.in:
```diff
@@ -1236,6 +1278,25 @@ ml <- function(...){
+
+         set fdef "function global:envmodule {"
+         append fdef {
+   param(
```
It would also be great to have support for `MODULES_RUN_QUARANTINE`, but if you have difficulties to understand how it works, no worries such support could be added later.

xdelaruelle reviewed (changes requested):
Many thanks for this contribution. That's great work.

In addition to the code review comment, main point is to add support in the installation testsuite (make testinstall). You will find it in `testsuite/install.00-init/`. This testsuite will check that `pwsh` correctly defines variable, function, etc with the commands generated by `modulecmd.tcl`

On the documentation side, things to add
* [How to initialize module on pwsh](https://github.com/cea-hpc/modules/blob/672e287ec50fb3c23988c23bcd07857a54546e46/doc/source/module.rst?plain=1#L134)
* [`pwsh` value returned by `module-info shell`](https://github.com/cea-hpc/modules/blob/672e287ec50fb3c23988c23bcd07857a54546e46/doc/source/modulefile.rst?plain=1#L769)
* [`pwsh` value returned by `module-info shelltype`](https://github.com/cea-hpc/modules/blob/672e287ec50fb3c23988c23bcd07857a54546e46/doc/source/modulefile.rst?plain=1#L781)
* [update Shell support section](https://github.com/cea-hpc/modules/blob/672e287ec50fb3c23988c23bcd07857a54546e46/doc/source/modulefile.rst?plain=1#L2083)

Support for `source-sh` or `sh-to-mod` commands may also be added (translate a pwsh script into a modulefile). But it may be the topic of another pull request.

ACSimon33 commented on .cirrus.yml:
```diff
@@ -97,7 +97,7 @@ task:
   reqinstall_script:
     - brew update
-    - brew install md5sha1sum expect dejagnu grep fish r tcl-tk gzip
+    - brew install md5sha1sum expect dejagnu grep fish r tcl-tk gzip powershell
```
Does the testsuite actually use the installed shells in some way? I was suprised that the linux tests succeeded even without pwsh installed.

ACSimon33 replied:
Ah, I see now that `make testinstall` is actually testing all the installed shells

xdelaruelle replied:
`pwsh` should be added to the `foreach shell` loop in `testsuite/install.00-init/005-init_ts.exp`

ACSimon33 commented on tcl/envmngt.tcl.in:
```diff
@@ -1236,6 +1278,25 @@ ml <- function(...){
+            reportErrorAndExit {No autoinit mode available for 'pwsh' shell on Windows}
+         }
+
+         set fdef "function global:envmodule {"
```
Return codes of functions are a bit weird in powershell. You can either use `Write-Error` or `throw` in powershell to get the calling shell to recognize that the function failed. I can try to capture the stderr stream from the modulecmd.tcl call and reprint the message via one of the two options. The only problem is that this will always add something to the error message:

`Write-Error`:
```
envmodule: ERROR: Invalid command 'foo'
  Try 'module --help' for more information.
```

`throw`:
```
Exception: ERROR: Invalid command 'foo'
  Try 'module --help' for more information.
```

So we have to account for that in the test suite. I'd vote for the exception since that could be useful if the user want's to catch that via `try {} catch {}`

ACSimon33 replied:
As an alternative, there is the `LastExitCode` variable which we could manually set inside the function. This variable is pretty much independent of the actual execution status `$?`. `$?` is only false if we exit the function via `Write-Error` or `throw`. One place where we don't want that is for example `module is-loaded ...`. If the module in question isn't loaded, I'd say we don't want to throw an exception but rather set the `LastExitCode` or actually return a bool

xdelaruelle replied:
On other shells, it returns `_mlstatus` which acts as a bool. I agree that it should not throw an error on query operations like `is-loaded`.

ACSimon33 replied:
We could always return `_mlstatus` but that means that each command prints either `True`, `False`, or whatever text is stored in it to the console. Powershell acts more like a language like Python than like "normal" shells. I'll see if I can find a better solution.

xdelaruelle commented on .cirrus.yml:
```diff
@@ -135,6 +135,11 @@ task:
     # requirements for source-sh extra tests
     apt-get -y install lsof
+    # install powershell
+    wget https://github.com/PowerShell/PowerShell/releases/download/v7.4.5/powershell_7.4.5-1.deb_amd64.deb
```
please use `curl` rather `wget` to download package

xdelaruelle commented on tcl/envmngt.tcl.in:
```diff
@@ -1236,6 +1282,29 @@ ml <- function(...){
+         if {[getState is_win]} {
+            reportErrorAndExit {
+               No autoinit mode available for 'pwsh' shell on Windows}
+         }
```
You may test coverage of this error in `testsuite/modules.00-init/120-siteconfig.exp`, within section where `TESTSUITE_ENABLE_SITECONFIG_WINDOWS_PLATFORM` variable is set

xdelaruelle commented on tcl/envmngt.tcl.in:
```diff
@@ -1236,6 +1282,29 @@ ml <- function(...){
+      pwsh {
+         if {[getState is_win]} {
+            reportErrorAndExit {
+               No autoinit mode available for 'pwsh' shell on Windows}
```
Please rewrite the following way to avoid trailing space in error message string:

```tcl
            reportErrorAndExit {No autoinit mode available for 'pwsh' shell\   
               on Windows}
```

ACSimon33 commented:
@xdelaruelle I've added pwsh to the othlang_list in the installation tests because it often doesn't behave like a normal shell. Also, it was easier to write the `install_test_pwsh` handling the custom module command (`envmodule`). I fixed most of the things it was complaining about. 
One issue is the support for `MODULES_REDIRECT_OUTPUT`. When we use exceptions as a way to signal an error, there is no way to redirect the message to stdout. And if we write the error message directly to stdout, we have no way of signaling that the command failed. We still can set the `LastExitCode`, but that is not really the current status code of the shell because it is normally only set by native applications and not pwsh functions/commands.

ACSimon33 commented:
@xdelaruelle I think I addressed everything now except the support for `sh-to-mod`. I don't see this as an urgent feature because most environment scripts on Windows are still batch files. This can be dealt with in a follow-up PR. 
The last this that we might want to change is the INSTALL.bat script on Windows. The problem is that it puts the module installation directory into the system's PATH variable. This interferes with the `module` keyword in pwsh. I'd say we just remove that from the INSTALL script. CMD users can just source the `init/cmd.cmd`, which will set the PATH for them.

xdelaruelle commented:
Many thanks Simon for this deep work.

I will take over if I spot some finishing changes to do.

I will test everything on my side.

xdelaruelle commented:
@ACSimon33 module quarantine tests are failing so I suspect there is an issue with the code of this feature. Could you please look at it to fix it. If this is too hard or time-consuming for you, I may remove the feature for the moment (could be added later on).

Beware that I have reworked the commits of your branch, so you may pull these changes first, prior to push new commits.

ACSimon33 commented:
> @xdelaruelle […] The last thing that we might want to change is the INSTALL.bat script on Windows. The problem is that it puts the module installation directory into the system's PATH variable. This interferes with the `module` keyword in pwsh. I'd say we just remove that from the INSTALL script. CMD users can just source the `init/cmd.cmd`, which will set the PATH for them.

@xdelaruelle Is it ok if we remove the environment variable from the INSTALL.bat script? Alternatively we could create a new install script for pwsh which just copies the files and doesn’t set the env variable. That way nothing would change for CMD users.

xdelaruelle commented:
> > @xdelaruelle […] The last thing that we might want to change is the INSTALL.bat script on Windows. The problem is that it puts the module installation directory into the system's PATH variable. This interferes with the `module` keyword in pwsh. I'd say we just remove that from the INSTALL script. CMD users can just source the `init/cmd.cmd`, which will set the PATH for them.
> 
> @xdelaruelle Is it ok if we remove the environment variable from the INSTALL.bat script? Alternatively we could create a new install script for pwsh which just copies the files and doesn’t set the env variable. That way nothing would change for CMD users.

@ACSimon33 I prefer no change for CMD users. Could you propose specific INSTALL/TESTINSTALL/UNINSTALL scripts for pwsh?

Subsidiary question: is there a way to automatically setup `envmodule` when `pswh` starts? Like done on `*sh` shells with profile script.

ACSimon33 commented:
> @ACSimon33 I prefer no change for CMD users. Could you propose specific INSTALL/TESTINSTALL/UNINSTALL scripts for pwsh?

Alright, I will create the corresponding pwsh scripts.

> Subsidiary question: is there a way to automatically setup `envmodule` when `pswh` starts? Like done on `*sh` shells with profile script.

The path to the profile script for pwsh is stored in the `$PROFILE` variable. On Windows this is `$env:HOME\Documents\PowerShell\Microsoft.PowerShell_profile.ps1` and on Unix it is at `$env:HOME/.config/powershell/Microsoft.PowerShell_profile.ps1` (https://learn.microsoft.com/en-us/powershell/scripting/learn/shell/creating-profiles?view=powershell-7.4#how-to-create-your-personal-profile)

xdelaruelle commented:
Many thanks again @ACSimon33 for your deep work on this. Everything seems ready. I will merge the PR as soon as the CI completes.

TaiXeflar commented:
Hi, 
I just wondering why not just using name `modules` to solve conflicted name `module` with pwsh?
the command it just looks like:
```
 modules av
 modules load ... ... ...
 modules purge
```

xdelaruelle commented:
@TaiXeflar `envmodule` is already the name used on Ruby for the same kind of name conflict than for PowerShell. I also find it less confusing for people not knowing Modules and looking at script using its function: just an extra `s` may confuse them with the reserved keyword.
